Speed up building Docker containers, upgrade container to Ubuntu 22.04, test Clang 14

nilsvu opened the pull request:
## Proposed changes

- Installing CMake from source takes a long time (2.5h when emulating ARM on CI!), so download prebuilt binaries instead.
- Doxygen takes another 1.5h to compile when emulating ARM on CI. Instead, it is now cross-compiled for ARM on the host x86 machine, then copied over to the ARM container. (took me a while to figure out, but not as long as I expected given how complicated cross-compiling can be. The result is surprisingly straightforward, and can be used to cross-compile other packages in the future as well if needed.)
- Upgrade container to Ubuntu 22.04 and install packages through apt if available. Those are already available for x86 and ARM so we don't have to build them from source (most importantly HDF5, which takes another 1.5h to compile when emulating ARM on CI).
- Test clang 14 on CI because it's in the container now, so this was quick to add.

These speedups help build the ARM container on CI, which has to emulate ARM. Here's a CI run: https://github.com/nilsvu/spectre/actions/runs/6129266583

The other dependencies that take a long time are:

- [x] ccache (Ubuntu 22.04 has v4.5.1, which is probably enough)
- [x] Doxygen (need v1.9.3+, which is available only on Ubuntu 23.04)
- [x] HDF5 (Ubuntu 22.04 has v1.10.7, which is probably enough)
- [ ] Charm++ (twice: multicore and MPI version)

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Change ordering in Dockerfile to improve caching

Avoids unnecessarily re-running the these steps.

nilsvu committed:
Install prebuilt CMake in container

Avoids building it from source, which can take a long time, especially
when emulating another architecture.
For CMake 3.18 no ARM binary is available, so install it only in
x86 CI container.

nilsvu committed:
Don't change `WORKDIR` in container

Make sure it's `/work` on container entry (was `/work/texlive` before).

nilsvu committed:
Upgrade container to Ubuntu 22.04

nilsvu committed:
Cross-compile some expensive builds in container

Compile these on the host system for the target architecture. This is
much faster than emulation. Fairly straightforward for simple
binaries like Doxygen and ccache and saves
a lot of time.

nilsvu committed:
Remove unused ruby/coveralls from container

nilsvu committed:
Update devcontainer file

nilsvu committed:
Test clang 14 on CI

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -342,6 +339,12 @@ RUN wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - \
     && apt-get update -y && apt-get install -y clang-13
 
+# Install minimum required CMake version so that we test compatibility with it
+RUN wget https://github.com/Kitware/CMake/releases/download/v3.18.2/cmake-3.18.2-linux-${CMAKE_ARCH}.sh -O cmake-install.sh \
```
Sorry, what line above causes this to be skipped on ARM?

nilsvu replied:
This isn't skipped on ARM. Currently we don't build the CI container on ARM at all. This step (installing CMake v3.18) would be one of the steps that fail (because Kitware doesn't publish a v3.18 ARM binary). I haven't actually tried getting the CI container to build on ARM.

nilsdeppe replied:
Ah okay, thanks!

nilsvu replied:
I added a conditional to install this only on x86 and checked that the CI container builds on ARM. We still don't build it on CI or push it to Dockerhub though, because currently GitHub doesn't have any ARM runners.

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -304,19 +247,16 @@ RUN git clone --single-branch --branch v7.0.0 --depth 1 \
     && git apply /work/v7.0.0.patch \
-    && ./build LIBS multicore-linux-${CHARM_ARCH} gcc \
-      ${PARALLEL_MAKE_ARG} -g -O2 --build-shared \
+    && ./build LIBS multicore-linux-${CHARM_ARCH} clang \
```
sorry why switch to clang?

nilsvu replied:
I may have had a reason but don't remember, maybe just consistency with the MPI Charm build below. Prefer to drop this change?

nilsdeppe replied:
I think I intentionally wanted different builds just to make sure we could do both GCC & clang with charm++ :shrug: So slight preference to drop

nilsvu replied:
ok, dropped 👍

nilsdeppe reviewed (changes requested):
LGTM, just a couple clarifying questions. Happy to approve once I understand :)

nilsvu commented:
Made the small change above and reduced the number of clang builds on CI. The clang-14 build isn't marked as "required" so we can merge PRs fine without rebasing.

nilsvu committed:
Upgrade xsimd in container

nilsvu committed:
Remove cppcheck

Unused for a long time. Also there's support for cppcheck in
CMake now, so if we want to use it again we don't need these
CMake files.

nilsvu committed:
Upgrade requirements-release.txt

Need support for Python 3.10, which is now the default
in the upgraded container.

nilsvu committed:
Fix some linking

nilsdeppe commented:
I assume I should be expecting the clang-14 failure, right? When do we want to merge this?

nilsvu committed:
Increase SwshDerivatives test timeout

This timed out on a couple of builds with the upgraded container.
Libsharp didn't change so I don't know what might cause this.

nilsvu committed:
Move ASAN CI tests to clang-14

clang-11 has an issue with ASAN and latest glibc:
https://github.com/llvm/llvm-project/issues/55785
So just use clang-14 for the ASAN build.

nilsvu committed:
Test Python + CMake backwards compatibility on CI

These were removed when the container used the minimum required
versions of Python and CMake by default, but now it uses newer
versions.

nilsvu commented:
Yes, clang-14 will fail until the upgraded container is pushed. I'm testing CI in the container on my fork. I'll message you when it's ready.

nilsvu committed:
Test install target on CI only for one build

Fixes issues with running out of disk space, which also happens
on current develop.

nilsvu committed:
Use libbacktrace on CI

knelli2 commented:
I was able to build the `dev` and `ci` containers on my local linux machine. I know @nilsvu is still debugging an issue on running CI right now, so we should wait on that. @nilsvu do you have a preference on if #5447 should me merged before or after this PR since it'll create conflicts?

nilsvu commented:
Merge https://github.com/sxs-collaboration/spectre/pull/5447 first because that's ready to go.

nilsvu commented:
Ok this is ready for review now. I think I worked out all the issues that came from upgrading the container to Ubuntu 22.04.

Plan:

- [x] Review
- [x] Push upgraded container
- [x] Run CI on this PR, then merge
- [ ] Rebase PRs

knelli2 commented:
@nilsvu Can you paste a link to the CI on your fork that passed with the new container?

nilsvu commented:
https://github.com/nilsvu/spectre/actions/runs/6181442942

knelli2 commented:
The Burgers executable in gcc-10 Debug failed the execute input file test with a seg fault. Do you know if that's from the container?

nilsvu commented:
No that's https://github.com/sxs-collaboration/spectre/issues/5384

knelli2 commented:
New container being built [here](https://github.com/nilsvu/spectre/actions/runs/6188302743/job/16800016645). It will push to `sxscollaboration/spectre` when it is finished (if all goes according to plan)
